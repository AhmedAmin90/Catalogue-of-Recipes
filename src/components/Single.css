.Single {
    display: flex;
    flex-direction: column;
    width: 50%;
    margin: 0 auto;
}

.Single-title {
    color: #f52;
    font-size: 20px;
    text-transform: capitalize;
    font-style: italic;
}

.Single-img {
   border-radius: 10px;
}

.Single-ingredient{
    list-style:inside;
    padding: 10px;
    background-color: #215330;
    opacity: 0.8;
}

.Single-ingredient > li {
    font-size: 15px;
    color: #FCC93C;
}