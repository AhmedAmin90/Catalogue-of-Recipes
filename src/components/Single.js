/* eslint-disable consistent-return, array-callback-return */
import React, { useState, useEffect } from 'react';
import PropTypes from 'prop-types';
import axios from 'axios';
import { NavLink } from 'react-router-dom';
import './Single.css';

const Single = ({ foodData }) => {
  const selectedId = foodData.match.params.id;

  const [meal, setMeal] = useState({
    id: selectedId,
    name: '',
    ingredients: [],
    image: '',
  });
  const [axiosRes, setAxiosRes] = useState('');

  useEffect(() => {
    const cancelToken = axios.CancelToken;
    const source = cancelToken.source();
    setAxiosRes(axiosRes);
    async function getDetails() {
      try {
        const res = await axios.get(`https://www.themealdb.com/api/json/v1/1/lookup.php?i=${selectedId}`, {
          cancelToken: source.token,
        });
        const data = await res.data.meals[0];
        const ingredientsArr = [];
        const dataArr = Object.keys(data);
        for (let i = 0; i < dataArr.length; i += 1) {
          if (dataArr[i].indexOf('Ingredient') !== -1) {
            ingredientsArr.push(data[dataArr[i]]);
          }
        }
        setMeal((pre) => ({
          ...pre,
          name: data.strMeal,
          ingredients: ingredientsArr,
          image: data.strMealThumb,
        }));
      } catch (err) {
        return err;
      }
    }

    getDetails();
    return () => {
      source.cancel('axios request cancelled');
    };
  }, []);

  const ingredientList = meal.ingredients.map((ing) => {
    if (ing !== '' && ing !== null) {
      return (
        <li key={`${ing}${Math.random() * 10}`}>

          {ing}

        </li>
      );
    }
  });
  return (
    <div className="Single">
      <nav>
        <NavLink to="/"> Back To Home </NavLink>
      </nav>

      <div className="Single-container">
        <h1 className="Single-title">{meal.name}</h1>
        <img className="Single-img" src={meal.image} alt={meal.name} />
        <h1 className="Single-title">List of Ingredients: </h1>
        <ul className="Single-ingredient">
          {ingredientList}
        </ul>
      </div>
    </div>
  );
};

Single.defaultProps = {
  foodData: {
    history: {},
    location: {},
    match: { params: { id: '1' } },
  },
};

Single.propTypes = {
  foodData: PropTypes.instanceOf(Object),
};

export default Single;
