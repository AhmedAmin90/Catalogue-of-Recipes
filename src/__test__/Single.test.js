import { render, screen, act } from '@testing-library/react';
import '@testing-library/jest-dom';
import '@testing-library/jest-dom/extend-expect';
import { create } from 'react-test-renderer';
import { Provider } from 'react-redux';
import {
  BrowserRouter as Router,
} from 'react-router-dom';
import store from '../index';
import Single from '../components/Single';

test('renders correctly', () => {
  const tree = create(<Provider store={store}><Router><Single /></Router></Provider>).toJSON();
  expect(tree).toMatchSnapshot();
});

describe('Test', () => {
  it('loading', async () => {
    const routeComponentPropsMock = {
      history: {},
      location: {},
      match: { params: { id: '52794' } },
    };

    await act(async () => render(
      <Provider store={store}>
        <Router>
          <Single foodData={routeComponentPropsMock} />
          ;
        </Router>
      </Provider>,
    ));
    expect(screen.getByText('List of Ingredients:')).toBeInTheDocument();
  });
});
