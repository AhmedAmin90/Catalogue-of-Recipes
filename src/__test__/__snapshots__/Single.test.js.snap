// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  className="Single"
>
  <nav>
    <a
      aria-current="page"
      className="active"
      href="/"
      onClick={[Function]}
      style={Object {}}
    >
       Back To Home 
    </a>
  </nav>
  <div
    className="Single-container"
  >
    <h1
      className="Single-title"
    >
      
    </h1>
    <img
      alt=""
      className="Single-img"
      src=""
    />
    <h1
      className="Single-title"
    >
      List of Ingredients: 
    </h1>
    <ul
      className="Single-ingredient"
    />
  </div>
</div>
`;
